mautic/mautic
# Move removal of duplicate indexes from fixture to Doctrine event listener

| Q                                      | A
| -------------------------------------- | ---
| Bug fix? (use the a.b branch)          | [ ]
| New feature/enhancement? (use the a.x branch)      | [x]
| Deprecations?                          | [ ]
| BC breaks? (use the c.x branch)        | [ ]
| Automated tests included?              | [ ] 
| Related user documentation PR URL      | mautic/mautic-documentation#... 
| Related developer documentation PR URL | mautic/developer-documentation#... 
| Issue(s) addressed                     | Fixes #... 



#### Description:

When testing and working on the PHP 8 support (see https://github.com/mautic/mautic/pull/9969#issuecomment-1020322075), I noticed there are some fixtures that contain [DDL queries](https://www.geeksforgeeks.org/sql-ddl-dql-dml-dcl-tcl-commands/).

As Doctrine processes fixtures as 1 transaction that can be rolled back, and DDL queries force a commit when executed,
this results in an error when Doctrine tries to commit the transaction again.

in PHP < 8, this error was silently ignored, but in PHP >= 8 this error results in a ` There is no active transaction` PDOException.

One of the fixtures removes duplicate indexes on the `email_assets_xref` and `email_list_xref` tables.

This was only a partial (and with PHP 8 a breaking) fix, as there are multiple tables that have these duplicate indexes.  

Single column indexes that are the leftmost column of another index are obsolete.
That leftmost column is available to look up rows, so there is no need for an extra index.

See https://dev.mysql.com/doc/refman/5.7/en/multiple-column-indexes.html

That extra index is automatically created by Doctrine (e.g. for many to many join tables), and may be fixed in the future.
Luckily there are events we can subscribe to, so the duplicate indexes can be removed.

This PR adds the cleanup functionality for all duplicate indexes.
As side effect, all obsolete for many to many tables are deleted too

TO be able to test the impact, I used [`Percona tools duplicate key checker`](https://www.percona.com/doc/percona-toolkit/LATEST/pt-duplicate-key-checker.html)

https://github.com/percona/percona-toolkit/blob/3.x/bin/pt-duplicate-key-checker

**BEFORE**

```
# ########################################################################
# Summary of indexes
# ########################################################################

# Size Duplicate Indexes   8810
# Total Duplicate Indexes  40
# Total Indexes            618
```

**AFTER**

```
# ########################################################################
# Summary of indexes
# ########################################################################

# Size Duplicate Indexes   9497
# Total Duplicate Indexes  24
# Total Indexes            602
```



#### Steps to test this PR:


1. Open this PR on Gitpod or pull down for testing locally (see docs on testing PRs [here](https://contribute.mautic.org/contributing-to-mautic/tester))
2. install Mautic, and check if the many to many tables (you can recognise them by the `_xref` suffix) don't have an additional index for the first column in the primary key




<a href="https://gitpod.io/#https://github.com/mautic/mautic/pull/11058"><img src="https://gitpod.io/button/open-in-gitpod.svg"/></a>
